blockchain/message_pool/src/msgpool.rs

```diff
@@ -0,0 +1,1033 @@
+    /// at. Return ActorState or Error depending on whether or not ActorState is found
+    fn state_get_actor(&self, addr: &Address, ts: &Tipset) -> Result<ActorState, Error>;
+    /// Return the signed messages for given blockheader
+    fn messages_for_block(
+        &self,
+        h: &BlockHeader,
+    ) -> Result<(Vec<UnsignedMessage>, Vec<SignedMessage>), Error>;
+    /// Return all messages for a tipset
+    fn messages_for_tipset(&self, h: &Tipset) -> Result<Vec<UnsignedMessage>, Error>;
+    /// Return a tipset given the tipset keys from the ChainStore
+    fn load_tipset(&self, tsk: &TipsetKeys) -> Result<Tipset, Error>;
+}
+
+/// This is the mpool provider struct that will let us access and add messages to messagepool.
+/// future
+pub struct MpoolProvider<DB> {
+    cs: ChainStore<DB>,
+}
+
+impl<'db, DB> MpoolProvider<DB>
+where
+    DB: BlockStore,
+{
+    pub fn new(cs: ChainStore<DB>) -> Self
+    where
+        DB: BlockStore,
+    {
+        MpoolProvider { cs }
+    }
+}
+
+impl<DB> Provider for MpoolProvider<DB>
+where
+    DB: BlockStore,
+{
+    fn subscribe_head_changes(&mut self) -> Subscriber<Arc<Tipset>> {
+        self.cs.subscribe()
+    }
+
+    fn get_heaviest_tipset(&mut self) -> Option<Tipset> {
+        let ts = self.cs.heaviest_tipset()?;
+        Some(ts.as_ref().clone())
+    }
+
+    fn put_message(&self, msg: &SignedMessage) -> Result<Cid, Error> {
+        let cid = self
+            .cs
+            .db
+            .put(msg, Blake2b256)
+            .map_err(|err| Error::Other(err.to_string()))?;
+        Ok(cid)
+    }
+
+    fn state_get_actor(&self, addr: &Address, ts: &Tipset) -> Result<ActorState, Error> {
+        let state = StateTree::new_from_root(self.cs.db.as_ref(), ts.parent_state())
+            .map_err(Error::Other)?;
+        let actor = state.get_actor(addr).map_err(Error::Other)?;
+        actor.ok_or_else(|| Error::Other("No actor state".to_owned()))
+    }
+
+    fn messages_for_block(
+        &self,
+        h: &BlockHeader,
+    ) -> Result<(Vec<UnsignedMessage>, Vec<SignedMessage>), Error> {
+        chain::block_messages(self.cs.blockstore(), h).map_err(|err| err.into())
+    }
+
+    fn messages_for_tipset(&self, h: &Tipset) -> Result<Vec<UnsignedMessage>, Error> {
+        messages_for_tipset(self.cs.blockstore(), h).map_err(|err| err.into())
+    }
+
+    fn load_tipset(&self, tsk: &TipsetKeys) -> Result<Tipset, Error> {
+        self.cs.tipset_from_keys(tsk).map_err(|err| err.into())
+    }
+}
+
+/// This is the main MessagePool struct
+pub struct MessagePool<T: 'static> {
+    local_addrs: Vec<Address>,
+    pending: Arc<RwLock<HashMap<Address, MsgSet>>>,
+    pub cur_tipset: Arc<RwLock<Tipset>>,
+    api: Arc<RwLock<T>>,
+    pub min_gas_price: BigInt,
+    pub max_tx_pool_size: i64,
+    pub network_name: String,
+    bls_sig_cache: Arc<RwLock<LruCache<Cid, Signature>>>,
+    sig_val_cache: LruCache<Cid, ()>,
+    local_msgs: Vec<SignedMessage>,
```

Comment:
This should really be a hashmap, as it was before. There is no deduplication you are handling in here, and most interactions are less efficient because of how they are used. Can you give me a functional reason why this should be a Vector?